File: python/src/stacks_queues/queue_using_two_stacks.py
"""Queue implemented with two stacks (amortized O(1) push/pop)."""


class QueueUsingTwoStacks:
    def __init__(self) -> None:
        self._in: list[int] = []
        self._out: list[int] = []

    def push(self, x: int) -> None:
        self._in.append(x)

    def _pour(self) -> None:
        if not self._out:
            while self._in:
                self._out.append(self._in.pop())

    def pop(self) -> int:
        self._pour()
        return self._out.pop()

    def peek(self) -> int:
        self._pour()
        return self._out[-1]

    def empty(self) -> bool:
        return not self._in and not self._out

File: typescript/src/stacks_queues/queueUsingTwoStacks.ts
export class QueueUsingTwoStacks {
  private readonly inStack: number[] = [];
  private readonly outStack: number[] = [];

  push(x: number): void {
    this.inStack.push(x);
  }

  private pour(): void {
    if (this.outStack.length === 0) {
      while (this.inStack.length > 0) {
        this.outStack.push(this.inStack.pop()!);
      }
    }
  }

  pop(): number {
    this.pour();
    return this.outStack.pop()!;
  }

  peek(): number {
    this.pour();
    return this.outStack[this.outStack.length - 1]!;
  }

  empty(): boolean {
    return this.inStack.length === 0 && this.outStack.length === 0;
  }
}
